zerver/lib/idempotent_request.py

```diff
@@ -0,0 +1,180 @@
+from collections.abc import Callable
+from functools import wraps
+from typing import Any, ParamSpec, TypeVar, cast
+
+from django.db import connection, models, transaction
+from psycopg2.errors import LockNotAvailable
+from psycopg2.extras import DictCursor, DictRow
+from psycopg2.sql import SQL, Identifier, Placeholder
+
+from zerver.lib.exceptions import LockedError
+from zerver.models import IdempotentRequest, Realm, UserProfile
+
+WorkReturnT = TypeVar("WorkReturnT")
+ParamT = ParamSpec("ParamT")
+
+
+def do_idempotent_insert(
+    idempotency_key: str,
+    realm: Realm,
+    user: UserProfile,
+    destination_model: type[models.Model],
+    field_names: list[str],
+    column_to_value: dict[str, Any],
+) -> DictRow:
```

Comment:
This is still returning information which is tied to your underlying implementation.  Why doesn't this return `int`?  Why make callers dig in an extract `new_id`?  This also disagrees with the stated return type on line 34, below.

Reply:
Yes, I agree